libs/contrib/Data/Nat/Factor.idr

```diff
@@ -0,0 +1,333 @@
+    step (SearchArgs (S a) (S b) bLteA {bNonZero}) rec with (divMod (S a) (S b))
+        | Fraction (S a) (S b) q FZ prf =
+            let sbIsFactor = the (plus q (mult b q) = S a) $
+                    rewrite multCommutative b q in
+                    rewrite sym $ multRightSuccPlus q b in
+                    replace {P = \x => x = S a} (plusZeroRightNeutral (q * S b)) prf
+                skDividesA = CofactorExists q (lteTransitive bNonZero bLteA) sbIsFactor
+                skDividesB = selfFactor (S b)
+                greatest = \q', (CommonFactorExists q' _ qfb) => qfb
+            in
+            (S b ** MkGCD (CommonFactorExists (S b) skDividesA skDividesB) greatest)
+
+        | Fraction (S a) (S b) q (FS r) prf =
+                let rLtSb = lteSuccRight $ elemSmallerThanBound r
+                    qGt1 = the (LTE 1 q) $ case q of
+                        Z => absurd . notLteAndGt (S $ finToNat r) b (elemSmallerThanBound r) $
+                            replace {P = LTE (S b)} (sym prf) bLteA
+                        (S k) => LTESucc LTEZero
+                    smaller = the (LTE (S (S (plus b (S (finToNat r))))) (S (plus a (S b)))) $
+                        rewrite plusCommutative a (S b) in
+                        LTESucc . LTESucc . addLteLeft . fromLteSucc $ lteTransitive (elemSmallerThanBound $ FS r) bLteA
+                    (f ** MkGCD (CommonFactorExists f prfSb prfRem) greatestSbSr) =
+                        rec (SearchArgs (S b) (S $ finToNat r) rLtSb) smaller
+                    prfSa = the (Factor (S a) f) $
+                        rewrite sym prf in
+                        rewrite multCommutative q (S b) in
+                        plusFactor (multNAlsoFactor prfSb q) prfRem
+                    greatest = \q', (CommonFactorExists q' qfa qfb) =>
+                        let sbfqSb =
+                                the (Factor (q * S b) (S b)) $
+                                rewrite multCommutative q (S b) in
+                                multFactor (S b) q
+                            rightPrf = minusFactor {a = q * S b} {b = S (finToNat r)}
+                                (rewrite prf in qfa)
+                                (factorTransitive sbfqSb qfb)
+                        in
+                        greatestSbSr q' (CommonFactorExists q' qfb rightPrf)
+                in
+                (f ** MkGCD (CommonFactorExists f prfSa prfSb) greatest)
+
+||| An implementation of Euclidean Algorithm for computing greatest common
+||| divisors. It is proven correct and total; returns a proof that computed
+||| number actually IS the GCD. Unfortunately it's very slow, so improvements
+||| in terms of efficiency would be welcome.
+export
+gcd : (a, b : Nat) -> {auto aok : LTE 1 a} -> {auto bok : LTE 1 b} -> (f : Nat ** GCD a b f)
+gcd Z _ {aok} = absurd aok
+gcd _ Z {bok} = absurd bok
+gcd (S a) (S b) with (cmp (S a) (S b))
+    gcd (S (b + S d)) (S b) | CmpGT d =
```

Comment:
One way to handle Euclid's algorithm is to arrange it so that the first argument is the smaller one. If the first argument is larger, just flip the arguments around and proceed as usual, as in `gcd 27 18 = gcd 18 27 = 9`. That would simplify the logic (just keep the simpler of the two branches), although I think it would require an `assert_total`.

This is discussed at the very beginning of Knuth's _The Art of Computer Programming_.

Reply:
I believe it's exactly what I have done. If `a > b`, then the recursion is run with `(a, b)` pair, otherwise – with `(b, a)`. The `GCD.Search` type even contains a proof that `a >= b`. Totality of the function is implied by the use of `sizeInd` function from `WellFounded` module, which provides looping capability already proved to be total.

Reply:
Ah, okay, I'm still working through the details.